connection_pool.go

```diff
@@ -98,9 +98,28 @@ func (b *backend) getNewConnection() (*connection, error) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
+	// Attempt to determine if we're over quota, if so, first start by cleaning out any
+	// connections that are closed, then recheck. We assert that the quota is only meant
+	// to count against open/in-use connections, so I don't care about closed ones.
 	if b.counter >= b.conf.ConnectionLimit {
-		go b.debugHitMaxConnections()
-		return nil, nil
+		newConns := make([]*connection, 0, b.counter)
+		for _, conn := range b.conns {
+			if !conn.IsClosed() {
+				newConns = append(newConns, conn)
+			}
+		}
+
+		// If the lengths are the same, we truly hit max connections and there's nothing
+		// to do so return
+		if len(newConns) == b.counter {
+			go b.debugHitMaxConnections()
+			return nil, nil
+		}
+
+		// We eliminated one or more closed connections, use the new list and reset our
```

Comment:
I think I like the logic it as it is (more obviously correct and returns dead connections immediately).

If you wanted to get fancy, maybe note the time when the connection is cleaned up from the pool so Idle can log if it gets called on an already-debug-cleaned-up connection. The idea being that we can see if Idle is getting called 150 hours later or if it's getting called 10ms later. Then combined with naive logging here we can differentiate the cases:
1. Idle gets called really quickly after we clean up here.
2. Idle gets called a long time after we clean up here.
3. Idle never gets called (need full log though...).

But again maybe not worth doing.